mesop/server/state_session.py

```diff
@@ -80,12 +95,100 @@ def clear_stale_sessions(self):
       del self.cache[key]
 
 
+class FileStateSessionBackend(StateSessionBackend):
+  """Stores state on the filesystem.
+
+  This backend allows state to be shared across processes on the same server. If session
+  affinity is not available, then the only way to improve performance is to scale
+  vertically.
+
+  The main bottleneck is disk read/write performance. If using this backend, you should
+  load test against expected traffic and state size to see if this backend meets your
+  needs.
+  """
+
+  # Hardcode for now, but probably make adjustable in the future.
+  _SESSION_TTL_MINUTES = 10
+  # Rate in which we will attempt to clear stale sessions. This is done to
+  # to reduce concurrent attempts to delete stale sessions.
+  #
+  # Example: If the rate is set to 0.1, that means the clear state sessions function
+  # will run 1 in 10 requests or 10% of the time.
+  _SESSION_CLEAR_RATE = 0.1
+
+  def __init__(self, base_dir: Path):
+    if base_dir is None:
+      raise RuntimeError(
+        "Base dir must be set when using `FileStateSessionBackend`"
+      )
+    self.base_dir = base_dir
+    self.prefix = "session"
+
+  def _make_file_path(self, token: str) -> Path:
+    return self.base_dir / (self.prefix + token)
+
+  def restore(self, token: str, states: States):
+    """Gets the saved state from the given token and updates the given state."""
+    file_path = self._make_file_path(token)
+    try:
+      with open(file_path, "rb") as f:
+        states_data = msgpack.unpackb(f.read())
+      _deserialize_state(states, states_data)
+    except FileNotFoundError as e:
+      raise Exception("Token not found in state session backend.") from e
+    else:
+      try:
+        os.remove(file_path)
+      except FileNotFoundError as e:
+        logging.warning("Failed to delete file after restore - " + str(e))
+
+  def save(self, token: str, states: States):
+    """Saves state to the backend with the given token."""
+    with open(self._make_file_path(token), "wb") as f:
+      f.write(msgpack.packb(_serialize_state(states)))  # type: ignore
+
+  def clear_stale_sessions(self):
+    """Deletes unused state data."""
+    if random.random() > self._SESSION_CLEAR_RATE:
```

Comment:
It seems a little weird to use a probabilistic way of cleaning up. Why not use a time elapsed approach (e.g. if the last cleanup was more than 10 minutes ago, i.e. _SESSION_TTL_MINUTES, then do the cleanup)?

Reply:
Yes, I mainly did this in the case that there are multiple mesop processes running, so we'd need some kind of global state across the processes. So it was just easier to randomly clear the data.

The alternative I considered was writing a sort of lock file. So basically if the file exists and it's not X minutes old, then don't delete anything.

Let me know if you'd prefer the lock file approach or if you have a better idea.

Reply:
I like your idea of doing a quasi-lock file - it seems fairly straightforward and even if there's some weird race conditions (e.g. two processes see the file is > X minute and tries to do the cleanup at the same time, it doesn't seem that bad to me, it could cause contention in the filesystem / raise exceptions, but it should basically still do the clean-up).

I was thinking about you could do a proper lockfile but you would need to use the fcntl module for file locking on Unix-based systems, or the msvcrt module on Windows; and this could have tricky edge cases (e.g. a process creates a lock-file and then dies, i.e. a stale lock).

Reply:
Ok. I implemented the quasi-lock file for now. We can update to something more robust if necessary.